### PR TITLE
add sync_options

### PR DESCRIPTION
**Pull request by yuxuan-z19**
Add `sync_options` as suggested in [PR#14](https://github.com/jwbuurlage/Bulk/pull/14)
All edits are recorded in `CHANGELOG.md`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,52 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
...
+  };
+  ```
+
+  And now
+
+  ```cpp
+  void sync(sync_options option={}) {...}
+  ```
+
+  where `option` defaults to `{.clear_queue=true, .tag=""}` as mentioned above.
```
Comment (jwbuurlage) [inline CHANGELOG.md]:
```suggestion
  where `option` defaults to `{.clear_queues=true, .tag=""}` as mentioned above.
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,52 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic
 Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [Unreleased] - 2023-02-21
+
+### Added
+
+- `world.sync()` now supports custom string tag
```
Comment (jwbuurlage) [inline CHANGELOG.md]:
```suggestion
- `world.sync()` now supports custom string tag that has no impact on behaviour, but can be used for bookkeeping.
```

File: include/bulk/world.hpp [hunk 3]
```diff
@@ -81,12 +99,27 @@ class world {
   /**
    * Perform a global barrier synchronization of the active processors and
    * resolves any outstanding communication. Messages previously received in
-   * queues are cleared for the next superstep, unless `clear_queues` is set
-   * to false. The function must be called by all processors. When some
+   * queues are cleared for the next superstep, 
+   * unless the member `clear_queues` of sync_option is set
```
Comment (jwbuurlage) [inline include/bulk/world.hpp]:
```suggestion
   * unless the member `clear_queues` of `sync_options` is set
```

File: include/bulk/world.hpp [hunk 4]
```diff
@@ -81,12 +99,27 @@ class world {
...
+   * to false. One may provide a tag to label the current call.
+   * The function must be called by all processors. When some
    * processors call `sync` while others call `barrier` at the same time,
    * behaviour is undefined.
    */
-  virtual void sync(bool clear_queues = true) = 0;
+  virtual void sync(sync_options option={}) = 0;
+
```
Comment (jwbuurlage) [inline include/bulk/world.hpp:111]:
```suggestion
void sync(bool clear_queues) { sync({.clear_queues = clear_queues}); }

```

to maintain backwards compatibility

File: test/communication.cpp [hunk 5]
```diff
@@ -609,5 +609,66 @@ void test_communication() {
...
+      world.sync({.clear_queues = false});
+      BULK_CHECK_ONCE(!q.empty(),
+                      "multiple messages arrived after not "
+                      "clearing queue");
+      BULK_CHECK_ONCE(world.sync_count("") == 2,
+                 "2 sync() with an empty tag were called since last reset");
+
+      q.clear();
+      
+      // set tag to random string : clear_queues == true, tag == random string
```
Comment (jwbuurlage) [inline test/communication.cpp]:
```suggestion
      // set tag to arbitrary string : clear_queues == true, tag == arbitrary string
```

Comment (jwbuurlage) [reply test/communication.cpp]: Let me know if you agree with these changes, then we can merge :)

Thanks for the quick fixes!

Comment (yuxuan-z19) [reply test/communication.cpp]: @jwbuurlage thanks for the corrections : )

**Comment by jwbuurlage** [comment]
Thanks  for your first contribution!

**Pull request merged**